pkg/events/queue/queue_mem_list.go

```diff
@@ -0,0 +1,114 @@
+package queue
+
+import (
+	"container/list"
+	"fmt"
+	"github.com/aquasecurity/tracee/pkg/external"
+	"sync"
+)
+
+type EventQueueMem struct {
+	cache                *list.List
+	mutex                sync.Mutex
+	maxAmountOfEvents    int // number of cached events
+	EventsCacheMemSizeMB int
+}
+
+func (q *EventQueueMem) String() string {
+	return fmt.Sprintf("In-Memory Event Queue (Size = %d MB)", q.EventsCacheMemSizeMB)
+}
+
+func (q *EventQueueMem) Setup() error {
+	q.maxAmountOfEvents = q.getQueueSizeInEvents()
+	q.cache = new(list.List)
+
+	return nil
+}
+
+// Enqueue pushes an event into the queue (may block until queue is available)
+func (q *EventQueueMem) Enqueue(evt *external.Event) {
+	if evt == nil {
+		return
+	}
+
+	// "spin" until cache length is smaller then maxAmountOfEvents (blocking)
+	for {
+		q.mutex.Lock() // can't be RO as PushBack will update the list
+		if q.cache.Len() > q.maxAmountOfEvents {
+			q.mutex.Unlock() // release mutex to allow concurrency
+			continue
+		}
+		break
+	}
+
+	q.cache.PushBack(*evt)
+	q.mutex.Unlock()
+	evt = nil
+}
+
+// Dequeue pops an event from the queue
+func (q *EventQueueMem) Dequeue(blocking bool) *external.Event {
+
+	// "spin" until cache length is bigger than 0
+	for {
+		q.mutex.Lock() // can't be RO as Remove() will update the list
+		if q.cache.Len() == 0 {
+			q.mutex.Unlock() // release mutex to allow concurrency
+			if !blocking {
```

Comment:
if `Dequeue()` is called with `blocking` input being `true`, and length of the queue is 0 then we will have a go routine running in a while true style, causing high cpu usage and locking operation. 
Moreover, even when we are not blocking, the outer loop (the one calling `Dequeue()`) will also run in a while true loop as long as len of queue = 0. 
I am not sure spinning is a good approach here.
What am I missing?
This can be solved with condition variables, in this way we do not have to spin - which is expensive. If we want to be blocked we can sleep on the condition variable (len of queue != 0)

In general, we can `RLock` when reading the queue length and `Lock` when calling Remove()

Reply:
We are not "spinning" (like spin locks) since mutexes allow the task to sleep/yield (so fair scheduler in Linux will be smart not to let this to step into other job if the preemption is needed).

Condition variables cannot be locked in a different "transaction", they would have to be locked while the "updates" are happening or else you always have a potential race in between the lock, the conditional variable and the data being changed.

My proposal to that concern is for us to sleep in this loop, and each time we loop we sleep a bit more until a threshold is reached. WDYT ?

This way if we have data to process we would be consuming CPU (obviously) but, if queue is empty for sometime, we would reduce CPU consumption.

Reply:
Okay as a feedback to this discussion, Michael and I just spoke on Teams and I was missing knowledge of the sync package functions he was proposing me to use. Now I'll change the loop to make use of wakeup like calls based on conditions (which was what I was missing).